expression-language.md

```diff
@@ -0,0 +1,375 @@
+```
+
+Because string literals can be either `''` or `""` delimited, in the former case, the `"` has to be escaped, while in
+the latter the `"` has to be escaped.
+
+### 2.3. Operators
+
+CESQL defines boolean unary and binary operators, arithmetic unary and binary operators and the `LIKE`, `IN`, `EXISTS`
+operators.
+
+```ebnf
+not-operator = "NOT"
+unary-logical-operator = not-operator
+binary-logic-operator = "AND" | "OR" | "XOR"
+
+unary-numeric-operator = "-"
+binary-comparison-operator = "=" | "!=" | "<>" | ">=" | "<=" | "<" | ">"
+binary-numeric-arithmetic-operator = "+" | "-" | "*" | "/" | "%"
+
+like-operator = "LIKE"
+exists-operator = "EXISTS"
+in-operator = "IN"
+
+unary-operation ::= (unary-numeric-operator | unary-logical-operator) expression
+
+binary-operation ::= expression (binary-comparison-operator | binary-numeric-arithmetic-operator | binary-logic-operator) expression
+
+like-operation ::= expression not-operator? like-operator string-literal
+
+exists-operation ::= exists-operator value-identifier
+
+set-expression ::= "(" expression ("," expression)* ")"
+in-operation ::= expression not-operator? in-operator set-expression
+```
+
+Note: the `set-expression` supports only compile-time set definition, hence the language doesn't include a _set_ type
+that can be manipulated at runtime.
+
+### 2.4. Functions invocation
+
+CESQL supports N arity function invocation:
+
+```ebnf
+char ::= [A-Z] | [a-z]
+parameter ::= expression
+
+function-identifier ::= char ( "_" | char )*
+
+parameter-list ::= parameter ("," parameter)*
+function-invocation ::= function-identifier "(" parameter-list? ")"
+```
+
+## 3. Language semantics
+
+### 3.1. Type System
+
+The type system contains 3 types:
+
+- _String_: Sequence of allowable Unicode characters
+- _Integer_: A signed 32-bit integer
+- _Boolean_: A boolean value of "true" or "false"
+
+The [types defined in the CloudEvents specification][ce-type-system] URI, URI Reference and String are represented as
+_String_.
+
+### 3.2. CloudEvent context identifiers and types
+
+Each CloudEvent context attribute and extension MUST be addressable from an expression using its identifier, as defined
+by the spec. For example, using `id` in an expression will address to the CloudEvent [id attribute][ce-id-attribute].
+
+Unless otherwise specified, every attribute and extension MUST be represented by the _String_ type. Through explicit
+casting, the user can convert the addressed value instances to _Integer_ and _Boolean_.
+
+When addressed an attribute not included in the input event, an empty _String_ MUST be assumed as value
+
+### 3.3. Errors
+
+Because every operator and function is total, an expression evalution flow is defined statically and cannot be modified
+by expected or unexpected errors. Nevertheless CESQL includes the concept of errors: when an expression is evaluated, in
+case an error arise, the evaluator collects a list of errors, referred in this spec as _error list_, which is then
```

Comment:
I think you throw or you don't. If you throw you get an error. If you don't, you get a result.

Reply:
I didn't wanted to place this constraint: the evaluation always ends, nevertheless an error happened or not. In other words, the evaluation can return:

* true, no error
* false, no error
* true, error
* false, error

This gives to the user the flexibility of deciding between a "strict" and a "soft" mode, and whether some errors may be more "critical" than others and treating such failures as `false`